From b58c700230430c744470e2fdedc73e2439a51346 Mon Sep 17 00:00:00 2001
From: CTCaer <ctcaer@gmail.com>
Date: Sat, 5 Mar 2022 03:30:44 +0000
Subject: [PATCH 02/20] fftools/libavformat: Enforce nvv4l2

This enforces NVV4L2 even if user requests another codec.
Additionally, it forces nvv4l2 to go through software codecs first to get context if needed.
---
 fftools/ffmpeg_opt.c | 19 +++++++++++++++++++
 fftools/ffplay.c     | 25 +++++++++++++++++++++++++
 libavformat/utils.c  | 13 +++++++++++++
 3 files changed, 57 insertions(+)

diff --git a/fftools/ffmpeg_opt.c b/fftools/ffmpeg_opt.c
index 807e783422..d6f28bda23 100644
--- a/fftools/ffmpeg_opt.c
+++ b/fftools/ffmpeg_opt.c
@@ -771,6 +771,25 @@ static const AVCodec *choose_decoder(OptionsContext *o, AVFormatContext *s, AVSt
     char *codec_name = NULL;
 
     MATCH_PER_STREAM_OPT(codec_names, str, codec_name, s, st);
+
+#if CONFIG_NVV4L2
+    /* Reset requested decoder in order to enforce NVV4L2 if possible. */
+    if (codec_name) {
+        if (strcmp(codec_name, "h264") == 0)
+            return avcodec_find_decoder(st->codecpar->codec_id);   
+        else if (strcmp(codec_name, "hevc") == 0)
+            return avcodec_find_decoder(st->codecpar->codec_id); 
+        else if (strcmp(codec_name, "mpeg2video") == 0)
+            return avcodec_find_decoder(st->codecpar->codec_id);
+        else if (strcmp(codec_name, "mpeg4") == 0)
+            return avcodec_find_decoder(st->codecpar->codec_id);
+        else if (strcmp(codec_name, "vp8") == 0)
+            return avcodec_find_decoder(st->codecpar->codec_id);
+        else if (strcmp(codec_name, "vp9") == 0 && st->codecpar->format != AV_PIX_FMT_YUV420P10)
+            return avcodec_find_decoder(st->codecpar->codec_id);
+    }
+#endif
+
     if (codec_name) {
         const AVCodec *codec = find_codec_or_die(codec_name, st->codecpar->codec_type, 0);
         st->codecpar->codec_id = codec->id;
diff --git a/fftools/ffplay.c b/fftools/ffplay.c
index e14c800b8f..e1c7b741b9 100644
--- a/fftools/ffplay.c
+++ b/fftools/ffplay.c
@@ -2601,6 +2601,31 @@ static int stream_component_open(VideoState *is, int stream_index)
         case AVMEDIA_TYPE_SUBTITLE: is->last_subtitle_stream = stream_index; forced_codec_name = subtitle_codec_name; break;
         case AVMEDIA_TYPE_VIDEO   : is->last_video_stream    = stream_index; forced_codec_name =    video_codec_name; break;
     }
+
+#if CONFIG_NVV4L2
+    /* Reset requested decoder in order to enforce NVV4L2 if possible. */
+    if (avctx->codec_type == AVMEDIA_TYPE_VIDEO && forced_codec_name) {
+        if (strcmp(forced_codec_name, "h264") == 0)
+            forced_codec_name = NULL;    
+        else if (strcmp(forced_codec_name, "hevc") == 0)
+            forced_codec_name = NULL; 
+        else if (strcmp(forced_codec_name, "mpeg2video") == 0)
+            forced_codec_name = NULL;
+        else if (strcmp(forced_codec_name, "mpeg4") == 0)
+            forced_codec_name = NULL;
+        else if (strcmp(forced_codec_name, "vp8") == 0)
+            forced_codec_name = NULL;
+        else if (strcmp(forced_codec_name, "vp9") == 0 &&
+                 avctx->pix_fmt != AV_PIX_FMT_YUV420P10) {
+            forced_codec_name = NULL;
+        }
+    }
+
+    /* NVV4L2 does not support VP9 with YUV420P10. */
+    if (!forced_codec_name && avctx->pix_fmt == AV_PIX_FMT_YUV420P10)
+        forced_codec_name = "vp9";
+#endif
+
     if (forced_codec_name)
         codec = avcodec_find_decoder_by_name(forced_codec_name);
     if (!codec) {
diff --git a/libavformat/utils.c b/libavformat/utils.c
index 1384b56771..a960f8265d 100644
--- a/libavformat/utils.c
+++ b/libavformat/utils.c
@@ -211,6 +211,19 @@ static const AVCodec *find_probe_decoder(AVFormatContext *s, const AVStream *st,
     if (codec_id == AV_CODEC_ID_H264)
         return avcodec_find_decoder_by_name("h264");
 #endif
+#if CONFIG_NVV4L2
+    /* NVV4L2 decoders depend on context init from base decoders */
+    if (codec_id == AV_CODEC_ID_HEVC)
+        return avcodec_find_decoder_by_name("hevc");
+    else if (codec_id == AV_CODEC_ID_MPEG2VIDEO)
+        return avcodec_find_decoder_by_name("mpeg2video");
+    else if (codec_id == AV_CODEC_ID_MPEG4)
+        return avcodec_find_decoder_by_name("mpeg4");
+    else if (codec_id == AV_CODEC_ID_VP8)
+        return avcodec_find_decoder_by_name("vp8"); 
+    else if (codec_id == AV_CODEC_ID_VP9)
+        return avcodec_find_decoder_by_name("vp9");
+#endif
 
     codec = find_decoder(s, st, codec_id);
     if (!codec)
-- 
2.25.1

